### ISSUE TITLE
Issues with querying larger graph using local triple store

### ISSUE THREAD

**Issue by SeFreid** [IBCNServices/pyRDF2Vec#186]
## ❓ Question

First of, thanks for providing a python implementation for rdf2vec it's been really useful to me so far.

I am using Python 3.9.16, pyrdf2vec version 0.2.3

I am running into some problems when trying to create an embedding of a somewhat larger graph (~1.1m entities). I have set up an Apache Fuseki triple store which I am giving pyRDF2Vec as input, this is working fine yet it still overflows the memory available to me (~190GB). As mentioned in previous issues I have tried to split the entities and using the get_walks() function on smaller chunks.  This is done using the following code snippet:

```
split_entities = np.array_split(entities, 10)

transformer = RDF2VecTransformer(
    Word2Vec(epochs=10),
    walkers=[RandomWalker(4, 50, n_jobs=38)],
    verbose=1
    )


walks = []
for entity in split_entities:
         walk = transformer.get_walks(kg, entity)
         walks.extend(walk)


transformer.fit(kg, entities)
embeddings,literals = transformer.transform(kg, entities)

```

It runs fine for the first set of entities, but afterwards I get the following error:

```
Traceback (most recent call last):
  File "/home/User/Documents/Tools/rtx-kg2/translator-lfs-artifacts/files/rdf2vec_remote.py", line 144, in <module>
    walk = transformer.get_walks(kg, entity)
  File "/home/User/miniconda3/envs/rdf2vec/lib/python3.9/site-packages/pyrdf2vec/rdf2vec.py", line 178, in get_walks
    walks += walker.extract(kg, entities, self.verbose)
  File "/home/User/miniconda3/envs/rdf2vec/lib/python3.9/site-packages/pyrdf2vec/walkers/walker.py", line 152, in extract
    kg._fill_hops(entities)
  File "/home/User/miniconda3/envs/rdf2vec/lib/python3.9/site-packages/pyrdf2vec/graphs/kg.py", line 440, in _fill_hops
    asyncio.run(self.connector.afetch(queries)),
  File "/home/User/miniconda3/envs/rdf2vec/lib/python3.9/site-packages/nest_asyncio.py", line 35, in run
    return loop.run_until_complete(task)
  File "/home/User/miniconda3/envs/rdf2vec/lib/python3.9/site-packages/nest_asyncio.py", line 90, in run_until_complete
    return f.result()
  File "/home/User/miniconda3/envs/rdf2vec/lib/python3.9/asyncio/futures.py", line 201, in result
    raise self._exception
  File "/home/User/miniconda3/envs/rdf2vec/lib/python3.9/asyncio/tasks.py", line 258, in __step
    result = coro.throw(exc)
  File "/home/User/miniconda3/envs/rdf2vec/lib/python3.9/site-packages/pyrdf2vec/connectors.py", line 103, in afetch
    return await asyncio.gather(*(self._fetch(query) for query in queries))
  File "/home/User/miniconda3/envs/rdf2vec/lib/python3.9/asyncio/tasks.py", line 328, in __wakeup
    future.result()
  File "/home/User/miniconda3/envs/rdf2vec/lib/python3.9/asyncio/tasks.py", line 256, in __step
    result = coro.send(None)
  File "/home/User/miniconda3/envs/rdf2vec/lib/python3.9/site-packages/pyrdf2vec/connectors.py", line 120, in _fetch
    async with self._asession.get(url, headers=self._headers) as res:
  File "/home/User/miniconda3/envs/rdf2vec/lib/python3.9/site-packages/aiohttp/client.py", line 1141, in __aenter__
    self._resp = await self._coro
  File "/home/User/miniconda3/envs/rdf2vec/lib/python3.9/site-packages/aiohttp/client.py", line 400, in _request
    raise RuntimeError("Session is closed")
RuntimeError: Session is closed
```

Any idea why this might be happening? The code runs just fine when I am not using a remote triple store, but I was hoping to speed it up a bit this way.

## ❓ Question

Another issue I came across was using the MultiProcessingRandomWalker described here: https://github.com/IBCNServices/pyRDF2Vec/wiki/Fast-generation-of-RDF2Vec-embeddings-with-a-SPARQL-endpoint

I copied the exact code and just replaced the RandomWalker with the MultiProcessingRandomWalker in my code

```
    transformer = RDF2VecTransformer(
    Word2Vec(epochs=10),
    walkers=[MultiProcessingRandomWalker(4, 50, n_jobs=38)],
    verbose=1
    )

```

Here I get the following error shortly after executing my code:

```
multiprocessing.pool.RemoteTraceback: 
"""
Traceback (most recent call last):
  File "/home/User/miniconda3/envs/rdf2vec/lib/python3.9/multiprocessing/pool.py", line 125, in worker
    result = (True, func(*args, **kwds))
  File "/home/User/Documents/Tools/rtx-kg2/translator-lfs-artifacts/files/rdf2vec_remote.py", line 22, in _proc
    kg, instance = t
ValueError: too many values to unpack (expected 2)
"""

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/User/Documents/Tools/rtx-kg2/translator-lfs-artifacts/files/rdf2vec_remote.py", line 146, in <module>
    embeddings, literals = transformer.fit_transform(kg, entities)
  File "/home/User/miniconda3/envs/rdf2vec/lib/python3.9/site-packages/pyrdf2vec/rdf2vec.py", line 143, in fit_transform
    self.fit(self.get_walks(kg, entities), is_update)
  File "/home/User/miniconda3/envs/rdf2vec/lib/python3.9/site-packages/pyrdf2vec/rdf2vec.py", line 178, in get_walks
    walks += walker.extract(kg, entities, self.verbose)
  File "/home/User/miniconda3/envs/rdf2vec/lib/python3.9/site-packages/pyrdf2vec/walkers/walker.py", line 155, in extract
    res = list(
  File "/home/User/miniconda3/envs/rdf2vec/lib/python3.9/site-packages/tqdm/std.py", line 1178, in __iter__
    for obj in iterable:
  File "/home/User/miniconda3/envs/rdf2vec/lib/python3.9/multiprocessing/pool.py", line 870, in next
    raise value
ValueError: too many values to unpack (expected 2)
Unclosed client session
client_session: <aiohttp.client.ClientSession object at 0x7ffd877dac10>
Unclosed connector
```

Do I have to modify the MultiProcessingRandomWalker code in any way? 

I hope it's alright, that I asked two questions in one issue. If not just tell me and I will open a separate issue for the MultiProcessingRandomWalker. Also sorry if I forgot any information, I tried to provide everything necessary. Thanks!

**Comment by bsteenwi** [comment]
Hi,

To answer your first question: this might be either an issue regarding the number of available threads your server can handle or an issue within the `aiohttp` package... We will further investigate your problem. Do you have the same issue when you run your code e.g. with n_jobs=4?

To answer your second question: I'm not sure if that code is still up to date.... but in general, it might be a good idea to take a look at [HDT](https://www.rdfhdt.org/) for large knowledge graphs.  We have already defined an HDTConnector and tested it in a separate branch. The creation of embeddings with HDT was very fast in our opinion.

If you are familiar with docker, you can easily transform your own KG into an HDT variant using the following command:
`docker run -it --rm -v "$(pwd)":/data rfdhdt/hdt-cpp rdf2hdt -f turtle /data/your_file.turtle /data/your_file.hdt`

Next, you can use our HDTConnector
```
class HDTConnector(Connector):
    def __attrs_post_init__(self):
        self.store = None

    @cachedmethod(operator.attrgetter("cache"))
    def fetch(self, query: str) -> Response:
        if self.store is None:
            self.store = rdflib_hdt.HDTStore(self.endpoint)
        try:
            res = self.store.hdt_document.search((query, None, None))[0]
            val = [{"p": {"value": r[1].toPython()},
                    "o": {"value": r[2].n3().split('"')[1]}} if isinstance(
                r[2], Literal) else {"p": {"value": r[1].toPython()},
                                     "o": {"value": r[2].toPython()}} for r
                   in res]
            return {"results": {"bindings": val}}
        except Exception as e:
            return {"results": {"bindings": []}}


    def get_query(self, entity: str, preds: Optional[List[str]] = None) -> str:
        return URIRef(entity)
```

and you should be able to use this connector in your code:

```
connector = HDTConnector("your_file.hdt")
knowledge_graph = KG()
knowledge_graph.connector = connector
...
```

Keep in mind that not all of our functionality (literal support, reverse walks, etc.) will work with this approach.
It is mainly used to speed up the vanilla pyRDF2Vec embedding creation.

**Comment by SeFreid** [comment]
Hi,

thank you for the fast reply!

I checked again with n_job=4 and the same error occurs, I have 40 cores available so that should not be the problem. The Fuseki instance is also still running, so there should be no problem there.  I tried using stardog instead of fuseki but I just get the error
```
Traceback (most recent call last):
  File "/home/User/Documents/Tools/rtx-kg2/translator-lfs-artifacts/files/rdf2vec_remote.py", line 64, in <module>
    kg = KG(location="http://localhost:5820/rtx-kg2/query/", is_remote=True,
  File "<attrs generated init pyrdf2vec.graphs.kg.KG>", line 26, in __init__
  File "/home/Userl/miniconda3/envs/rdf2vec/lib/python3.9/site-packages/attr/_make.py", line 3096, in __call__
    v(inst, attr, value)
  File "/home/User/miniconda3/envs/rdf2vec/lib/python3.9/site-packages/pyrdf2vec/utils/validation.py", line 58, in _check_location
    raise ValueError(
ValueError: 'location' must be a valid URL (got http://localhost:5820/rtx-kg2/query)
```

Eventhough stardog is up and running and can be queried using the built-in query function for example:

```
stardog query execute rtx-kg2 "select * where { ?s ?p ?o }"

```

If there is anything else I could do in regards to that problem that could help, just let me know.

As for the second problem, sadly I am unable to use docker in my work environment and literal support is an essential feature for my project so I will probably not be able to use it. It does look very interesting though!

Again, thanks for the reply!

**Comment by bsteenwi** [comment]
For Stardog: I think you should provide `http://localhost:5820/rtx-kg2` as url instead of `http://localhost:5820/rtx-kg2/query` but I'm not sure...

@rememberYou 
We might have a problem with the aiohttp session within our connector.py.
It looks quite similar to this issue: [RuntimeError: Session is closed when using aiohttp session](https://stackoverflow.com/questions/61043291/runtimeerror-session-is-closed-when-using-aiohttp-session)

I'm not sure why we implemented it like this, and what the consequences are related to aiohttp.
Maybe you remember?

**Comment by rememberYou** [comment]
@SeFreid From what I can understand, you have this initial session issue because you are defining the `KG` object with `mul_req=True` which allow bundling of SPARQL queries by opening and closing several connections simultaneously. You should try to remove this flag (default behavior) if this is the case. I suppose that the `aiohttp` library must have a limit in the number of sessions which can be opened in parallel. The normal behavior is that a session is open and closed after the execution of each SPARQL query. There are probably some rooms of improvement in this regard. 

@bsteenwi In the past, we had to find optimization way to reduce the HTTP latency of SPARQL requests and the use of `aiohttp` made it possible to make asynchronous requests. However, we tested this flag for remote Knowledge Graph of moderate size.
